pycheribuild/projects/run_fvp.py

```diff
@@ -664,8 +664,32 @@ def add_hostbridge_params(*params):
             # real time since otherwise each second of countdown takes around 2 minutes:
             if self.fvp_project.fvp_revision >= (0, 11, 13):
                 fvp_args += ["-C", "board.rtc_clk_frequency=300"]
+
+            tcp_ports = []
+
+            # Expose to the real host all TCP ports exposed by the FVP
+            if self.ssh_port is not None:
+                tcp_ports += [self.ssh_port]
+                print(coloured(AnsiColour.green, "Listening for SSH connections on localhost:", self.ssh_port, sep=""))
```

Comment:
Hm, I don't like how early this gets printed now, feels like it might be easier for users to not notice?

Reply:
Though it depends how much; if it's just a couple of lines that's probably fine, if it's any more than that we probably want to pass in either a list of messages to print or a callback where we can print this

Reply:
```
Providing /home/jrtc4/cheri over SMB to the guest. Use `mkdir -p /srcroot && mount_smbfs -I 10.0.2.4 -N //10.0.2.4/source_root /srcroot` to mount it
Providing /home/jrtc4/cheri/build over SMB to the guest. Use `mkdir -p /buildroot && mount_smbfs -I 10.0.2.4 -N //10.0.2.4/build_root /buildroot` to mount it
Providing /home/jrtc4/cheri/output over SMB to the guest. Use `mkdir -p /outputroot && mount_smbfs -I 10.0.2.4 -N //10.0.2.4/output_root /outputroot` to mount it
Providing /home/jrtc4/cheri/output/rootfs-morello-purecap over SMB to the guest. Use `mkdir -p /rootfs && mount_smbfs -I 10.0.2.4 -N //10.0.2.4/rootfs /rootfs` to mount it

Listening for SSH connections on localhost:12824
About to run QEMU with image /home/jrtc4/cheri/output/cheribsd-morello-purecap.img and loader/kernel /home/jrtc4/cheri/output/rootfs-morello-purecap/boot/kernel/kernel
/home/jrtc4/cheri/output/morello-sdk/bin/qemu-system-morello -M virt,gic-version=3 -cpu morello -bios edk2-aarch64-code.fd -m 2048 -nographic -drive if=none,file=/home/jrtc4/cheri/output/cheribsd-morello-purecap.img,id=drv,format=raw -device virtio-blk-pci,drive=drv -device virtio-net-pci,netdev=net0 -netdev 'user,id=net0,smb=/home/jrtc4/cheri<<<source_root@ro:/home/jrtc4/cheri/build<<<build_root:/home/jrtc4/cheri/output<<<output_root@ro:/home/jrtc4/cheri/output/rootfs-morello-purecap<<<rootfs,hostfwd=tcp::12824-:22' -device virtio-rng-pci
Built for target 'run-morello-purecap' in 0.0022268295288085938 seconds
/home/jrtc4/cheri/output/morello-sdk/FVP_Morello/models/Linux64_GCC-6.4/FVP_Morello --version
Morello FVP version detected as (0, 11, 16)
/home/jrtc4/cheri/output/morello-sdk/FVP_Morello/models/Linux64_GCC-6.4/FVP_Morello -C board.virtio_net.enabled=1 -C board.virtio_net.hostbridge.userNetworking=true -C board.virtio_net.hostbridge.userNetPorts=15160=22 -C board.virtio_net.transport=legacy -C board.virtio_rng.transport=legacy -C board.virtioblockdevice.image_path=/home/jrtc4/cheri/output/cheribsd-morello-purecap.img -C displayController=0 -C board.virtio_rng.enabled=1 -C board.virtio_rng.seed=0 -C board.virtio_rng.generator=2 --data Morello_Top.css.scp.armcortexm7ct=/home/jrtc4/cheri/output/morello-sdk/firmware/morello-fvp/scp_romfw.bin@0x0 --data Morello_Top.css.mcp.armcortexm7ct=/home/jrtc4/cheri/output/morello-sdk/firmware/morello-fvp/mcp_romfw.bin@0x0 -C css.scp.armcortexm7ct.INITVTOR=0x0 -C css.mcp.armcortexm7ct.INITVTOR=0x0 --data /home/jrtc4/cheri/output/morello-sdk/firmware/morello-fvp/uefi.bin@0x14200000 -C soc.scp_qspi_loader.fname=/home/jrtc4/cheri/output/morello-sdk/firmware/morello-fvp/scp_ap_image.bin -C soc.mcp_qspi_loader.fname=/home/jrtc4/cheri/output/morello-sdk/firmware/morello-fvp/mcp_image.bin -C css.pl011_uart_ap.unbuffered_output=1 -C css.scp.CS_Counter.use_real_time=1 -C board.rtc_clk_frequency=300 -C board.terminal_uart0_board.start_telnet=0 -C board.terminal_uart0_board.quiet=1 -C board.terminal_uart1_board.start_telnet=0 -C board.terminal_uart1_board.quiet=1 -C css.mcp.terminal_uart0.start_telnet=0 -C css.mcp.terminal_uart0.quiet=1 -C css.mcp.terminal_uart1.start_telnet=0 -C css.mcp.terminal_uart1.quiet=1 -C css.scp.terminal_uart_aon.start_telnet=0 -C css.scp.terminal_uart_aon.quiet=1 -C css.terminal_sec_uart_ap.start_telnet=0 -C css.terminal_sec_uart_ap.quiet=1 -C css.terminal_uart1_ap.start_telnet=0 -C css.terminal_uart1_ap.quiet=1 -q -C disable_visualisation=true -C css.terminal_uart_ap.quiet=1 -C 'css.terminal_uart_ap.terminal_command=echo %port >&4'
Waiting for FVP to start...
Listening for SSH connections on localhost:15160
Connecting to the FVP using telnet. Press CTRL+] followed by q<ENTER> to exit telnet and kill the FVP.
telnet 127.0.0.1 5003
Built for target 'run-fvp-morello-purecap' in 0.08836507797241211 seconds
```
is what you currently get for both. I guess this turns the FVP into:
```
/home/jrtc4/cheri/output/morello-sdk/FVP_Morello/models/Linux64_GCC-6.4/FVP_Morello --version
Morello FVP version detected as (0, 11, 16)
Listening for SSH connections on localhost:15160
/home/jrtc4/cheri/output/morello-sdk/FVP_Morello/models/Linux64_GCC-6.4/FVP_Morello -C board.virtio_net.enabled=1 -C board.virtio_net.hostbridge.userNetworking=true -C board.virtio_net.hostbridge.userNetPorts=15160=22 -C board.virtio_net.transport=legacy -C board.virtio_rng.transport=legacy -C board.virtioblockdevice.image_path=/home/jrtc4/cheri/output/cheribsd-morello-purecap.img -C displayController=0 -C board.virtio_rng.enabled=1 -C board.virtio_rng.seed=0 -C board.virtio_rng.generator=2 --data Morello_Top.css.scp.armcortexm7ct=/home/jrtc4/cheri/output/morello-sdk/firmware/morello-fvp/scp_romfw.bin@0x0 --data Morello_Top.css.mcp.armcortexm7ct=/home/jrtc4/cheri/output/morello-sdk/firmware/morello-fvp/mcp_romfw.bin@0x0 -C css.scp.armcortexm7ct.INITVTOR=0x0 -C css.mcp.armcortexm7ct.INITVTOR=0x0 --data /home/jrtc4/cheri/output/morello-sdk/firmware/morello-fvp/uefi.bin@0x14200000 -C soc.scp_qspi_loader.fname=/home/jrtc4/cheri/output/morello-sdk/firmware/morello-fvp/scp_ap_image.bin -C soc.mcp_qspi_loader.fname=/home/jrtc4/cheri/output/morello-sdk/firmware/morello-fvp/mcp_image.bin -C css.pl011_uart_ap.unbuffered_output=1 -C css.scp.CS_Counter.use_real_time=1 -C board.rtc_clk_frequency=300 -C board.terminal_uart0_board.start_telnet=0 -C board.terminal_uart0_board.quiet=1 -C board.terminal_uart1_board.start_telnet=0 -C board.terminal_uart1_board.quiet=1 -C css.mcp.terminal_uart0.start_telnet=0 -C css.mcp.terminal_uart0.quiet=1 -C css.mcp.terminal_uart1.start_telnet=0 -C css.mcp.terminal_uart1.quiet=1 -C css.scp.terminal_uart_aon.start_telnet=0 -C css.scp.terminal_uart_aon.quiet=1 -C css.terminal_sec_uart_ap.start_telnet=0 -C css.terminal_sec_uart_ap.quiet=1 -C css.terminal_uart1_ap.start_telnet=0 -C css.terminal_uart1_ap.quiet=1 -q -C disable_visualisation=true -C css.terminal_uart_ap.quiet=1 -C 'css.terminal_uart_ap.terminal_command=echo %port >&4'
Waiting for FVP to start...
Connecting to the FVP using telnet. Press CTRL+] followed by q<ENTER> to exit telnet and kill the FVP.
telnet 127.0.0.1 5003
Built for target 'run-fvp-morello-purecap' in 0.08836507797241211 seconds
```
?

Which then mirrors QEMU, except the command line is really rather unwieldy. I suspect also that a non-headless run actually ends up printing the SSH message that early _anyway_, it's only the headless run that starts the FVP early? So maybe this is fine.

Reply:
Oh, we don't even print that message for non-headless runs currently :) oops